### PR TITLE
feat(tx): add EIP-1559 transaction with priority fee support

### PR DESCRIPTION
**Pull request by dimxy**
Add typed transactions support for type 1 with [EIP-2930](https://eips.ethereum.org/EIPS/eip-2930) access lists and type 2 of [EIP-1559](https://eips.ethereum.org/EIPS/eip-1559) fee per gas changes

### CODE REVIEW

**Comment by laruh** [comment]
@dimxy please tell which rust toolchain version did you use? 
As I can see in the project is used old config version `rust-toolchain` with just one word stable, and not `rust-toolchain.toml`. 
Well it uses default stable version (I have this `rustc 1.76.0 (07dca489a 2024-02-04)` stable) but still. 
I suggest to update it to `rust-toolchain.toml` and specify version, and in this way to commit cargo lock file.
Also its better to add `edition` version in cargo tomls, as without it cargo uses `2015` version by default.

As I understood you cant expect to format code in the whole project with `cargo fmt` command, as root `Cargo.toml` doesnt have `[workspace]` members. lib.rs is just empty file
image
Well we dont have to create workspace members.
You can format `ethkey` and `ethcore` folders running these scripts, as we still can use fmt command by itself.
```
find ./ethcore/transaction/src -type f -name '*.rs' -exec rustfmt {} +
find ./ethkey/src -type f -name '*.rs' -exec rustfmt {} +
```

But if we decide to have workspaces, then we will need to additionally write package in komodefi deps.
(ps Its just the info, i dont think we need it.)
```toml
[dependencies]
ethkey = { git = "https://github.com/dimxy/mm2-parity-ethereum.git", branch = "eip1559-support", package = "ethkey" }
ethcore-transaction = { git = "https://github.com/dimxy/mm2-parity-ethereum.git", branch = "eip1559-support", package = "ethcore-transaction" }
```

PS: yeah, I understand that `rust-toolchain` in dependency project doesn't affect the dependent project, and dependencies are compiled with dependent project own toolchain. Its is primarily for developers working with the project directly.


I have warning related to `https://github.com/paritytech/ring`that it was not used in the crate graph. Do you have the same?
image

**Comment by dimxy** [comment]
> @dimxy please tell which rust toolchain version did you use?

I believe I use the toolchain set by rust-toolchain.toml: "nightly-2022-10-29" (rustup confirms this)

**Comment by dimxy** [comment]
> @dimxy please tell which rust toolchain version did you use? 

stable-x86_64-apple-darwin (overridden by '/Users/dimxy/repo/mm2-parity-ethereum/rust-toolchain')

**Comment by laruh** [comment]
UPD: sorry, no need in `rust-toolchain.toml` as we are going to make this repo part of komodefi

**Comment by dimxy** [comment]
> UPD: sorry, no need in `rust-toolchain.toml` as we are going to make this repo part of komodefi

maybe, until we do not move it to our repo, it is worth to fix the toolchain (as this repo has not had any updates for a long time)

**Comment by laruh** [comment]
> > UPD: sorry, no need in `rust-toolchain.toml` as we are going to make this repo part of komodefi
> 
> maybe, until we do not move it to our repo, it is worth to fix the toolchain (as this repo has not had any updates for a long time)

Well if its ok for you. It didnt take much time for me to fix clippy warnings for 1.76.0 stable version locally. 
then you should add toolchain.toml with specific version in channel, also not bad to add `components = ["rustfmt", "clippy"]` and add edition versions in Cargo tomls.

As for `[workspace]` members in root Caro.toml its up to you. you can use scripts which I provided to format code in necessary folders.

**Comment by dimxy** [comment]
If I set it to our "nightly-2022-10-29" I begin to receive that 'patch ring was not used in the crate graph' warning. I guess I did not receive it myself because I had some older default toolchain picked by this lib 'stable' setting. I think this is not good just to use any 'stable' ver and better to fix it to some specific rust ver where this warning does not appear (it would be probably older than ours).

**Comment by laruh** [comment]
> If I set it to our "nightly-2022-10-29" I begin to receive that 'patch ring was not used in the crate graph' warning. I guess I did not receive it myself because I had some older default toolchain picked by this lib 'stable' setting. I think this is not good just to use any 'stable' ver and better to fix it to some specific rust ver where this warning does not appear (it would be probably older than ours).

I see it with 1.76.0 stable version. And sometimes after cargo clean didnt see it, then saw it again. I think it is just there and independent to toolchain version

**Comment by dimxy** [comment]
As I can tell cargo clippy complains about the used version of the 'ring' lib like it is different from the version in cargo.lock or just outdated. In fact this lib is fetched from the parity repo (also cargo.lock looks correct as it points to this repo too). I guess it's not good to just change it to a newer version so I suggest ignoring this warning for now.
Btw we have plans to switch to a different eth lib in future like alloy-rs whene it is ready.

**Comment by laruh** [comment]
> I guess it's not good to just change it to a newer version so I suggest ignoring this warning for now.
Btw we have plans to switch to a different eth lib in future like alloy-rs whene it is ready.

Yep, agree.

PS: commit the Cargo.lock please

File: ethcore/transaction/src/transaction/eip1559.rs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    /// Transfered value.
+    pub value: U256,
+    /// Transaction data.
+    pub data: Bytes,
+    /// Access list.
+    pub access_list: AccessList,
+}
+
+impl Eip1559Transaction {
+    const fn payload_length(&self) -> usize { 9 }
```
Comment (laruh) [inline ethcore/transaction/src/transaction/eip1559.rs]: could you rename this function? payload in many contexts usually refers to bytes if Im not mistaken,while here you show the number of transaction fields. `payload_length` name confuses a bit.

File: ethcore/transaction/src/transaction/eip1559.rs [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        s.append(&self.value);
+        s.append(&self.data);
+        s.append(&self.access_list);
+        s.append(&self.v);
+        s.append(&self.r);
+        s.append(&self.s);
+    }
+
+    pub fn standard_v(&self) -> u8 {
+        self.v.try_into().unwrap_or_default() // we should have parity 0 or 1  for v2 txns
```
Comment (laruh) [inline ethcore/transaction/src/transaction/eip1559.rs]: are you sure that `unwrap_or_default` wont have unexpected behavior?

Comment (dimxy) [reply ethcore/transaction/src/transaction/eip1559.rs]: added v validation in [be513ae](https://github.com/KomodoPlatform/mm2-parity-ethereum/pull/1/commits/be513ae838fd5b6c3c425478fbe986511bbdebf2) and [b6064ce](https://github.com/KomodoPlatform/mm2-parity-ethereum/pull/1/commits/b6064ceec79ebf0ff896889196e2991efee2d58d)

File: ethcore/transaction/src/transaction/eip1559.rs (hunk 1 above)
Comment (dimxy) [reply ethcore/transaction/src/transaction/eip1559.rs]: done in [be513ae](https://github.com/KomodoPlatform/mm2-parity-ethereum/pull/1/commits/be513ae838fd5b6c3c425478fbe986511bbdebf2)

File: ethkey/src/error.rs [hunk 3]
```diff
@@ -57,9 +57,9 @@ impl error::Error for Error {
 	}
 }
 
-impl Into<String> for Error {
-	fn into(self) -> String {
-		format!("{}", self)
+impl From<Error> for String {
+	fn from(err: Error) -> String {
+		format!("{}", err)
```
Comment (shamardy) [inline ethkey/src/error.rs]: I don't think this is needed at all since Error implements Display.

File: ethcore/transaction/src/transaction.rs [hunk 4]
```diff
@@ -16,596 +18,682 @@
...
 type BlockNumber = u64;
+type Bytes = Vec<u8>;
+
+#[derive(Clone, PartialEq)]
+pub enum TxType {
+    Legacy,
+    Type1 = 1,
+    Type2 = 2,
+    Invalid,
+}
```
Comment (shamardy) [inline ethcore/transaction/src/transaction.rs]: Can you please add doc comments about what each transaction type is?

File: ethcore/transaction/src/transaction/eip1559.rs [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+impl rlp::Decodable for UnverifiedEip1559Transaction {
+    fn decode(d: &Rlp) -> Result<Self, DecoderError> {
+        let unsigned = Eip1559Transaction::decode(d)?;
+        let hash = keccak(d.as_raw());
+        let offset = unsigned.payload_size();
+        if d.as_raw().is_empty() {
+            return Err(DecoderError::RlpIsTooShort);
+        }
```
Comment (shamardy) [inline ethcore/transaction/src/transaction/eip1559.rs]: Isn't this checked as part of `Eip1559Transaction::decode(d)?`?

File: ethcore/transaction/src/transaction/eip2930.rs [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+impl rlp::Decodable for UnverifiedEip2930Transaction {
+    fn decode(d: &Rlp) -> Result<Self, DecoderError> {
+        let unsigned = Eip2930Transaction::decode(d)?;
+        let hash = keccak(d.as_raw());
+        let offset = unsigned.payload_size();
+        if d.as_raw().is_empty() {
+            return Err(DecoderError::RlpIsTooShort);
+        }
```
Comment (shamardy) [inline ethcore/transaction/src/transaction/eip2930.rs]: Same here, this is checked as part of E`ip2930Transaction::decode(d)`

**Review by shamardy — DISMISSED** [review]
LGTM apart from a few nits!

File: ethcore/transaction/src/transaction/eip1559.rs (hunk 2 above)
Comment (laruh) [reply ethcore/transaction/src/transaction/eip1559.rs]: Thanks

**Review by laruh — DISMISSED** [review]
Thanks for the fixes, LGTM!

non blocking suggestion, please add some info in PR description. that additionally to eip-1559 tx the eip-2930 also was added. now project contains 3 versions of tx.

File: ethcore/transaction/src/transaction.rs (hunk 4 above)
Comment (dimxy) [reply ethcore/transaction/src/transaction.rs]: added in [5dbe940](https://github.com/KomodoPlatform/mm2-parity-ethereum/pull/1/commits/5dbe94081c38472bd733c15a1b4f614e6dcc54dd)

File: ethcore/transaction/src/transaction/eip2930.rs (hunk 6 above)
Comment (dimxy) [reply ethcore/transaction/src/transaction/eip2930.rs]: fixed in [5dbe940](https://github.com/KomodoPlatform/mm2-parity-ethereum/pull/1/commits/5dbe94081c38472bd733c15a1b4f614e6dcc54dd)

**Comment by dimxy** [comment]
> non blocking suggestion, please add some info in PR description. that additionally to eip-1559 tx the eip-2930 also was added. now project contains 3 versions of tx.

added

**Review by shamardy — APPROVED** [review]

**Pull request merged**